### ISSUE TITLE
Cosmos: Spatial types, functions and spatial indexes

### ISSUE THREAD

**Issue by AndriySvyryd** [aspnet/EntityFrameworkCore#17317]
Use GeoJSON which supports point, polygon and linestring

**Comment by xtellurian** [comment]
I was super stoked to use EF Core with dotnet 3 - but I've run into an issue with GeoJSON and the Cosmos DB database.

The GeoJSON spec says that coordinates must be stored as follows:
```json
{
"type": "Point",
"coordinates": [125.6, 10.1]
}
```

So I created a C# POCO class like so:

```cs
public class GeoLocation
    {
        [JsonProperty("type")]
        public string Type { get; set; }
        [JsonProperty("coordinates")]
        public double[] Coordinates { get; set; }
    }
```
However using this with EF Core + Cosmos throws the following:

```
Exception has occurred: CLR/System.InvalidOperationException
Exception thrown: 'System.InvalidOperationException' in System.Private.CoreLib.dll: 'The property 'GeoLocation.Coordinates' could not be mapped, because it is of type 'double[]' which is not a supported primitive type or a valid entity type. Either explicitly map this property, or ignore it using the '[NotMapped]' attribute or by using 'EntityTypeBuilder.Ignore' in 'OnModelCreating'.'
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelValidator.ValidatePropertyMapping(IModel model, IDiagnosticsLogger`1 logger)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelValidator.Validate(IModel model, IDiagnosticsLogger`1 logger)
   at Microsoft.EntityFrameworkCore.Cosmos.Internal.CosmosModelValidator.Validate(IModel model, IDiagnosticsLogger`1 logger)
```

The biggest pain point for us is that Azure Search [requires](https://docs.microsoft.com/en-us/rest/api/searchservice/supported-data-types) that coordinates be serialised as GEOJson, which is just not possible when using EF Core.

In conclusion, it is not possible to use Cosmos, EF Core, and Azure Search with Geo Spacial data :(

**Comment by bricelam** [comment]
@AndriySvyryd Any easy workarounds using [NetTopologySuite.IO.GeoJSON](https://github.com/NetTopologySuite/NetTopologySuite.IO.GeoJSON)?

``` csharp
var geoJsonString = new GeoJsonWriter().Write(point);
var point = new GeoJsonReader().Read<Point>(geoJsonString);
```

**Comment by AndriySvyryd** [comment]
Yes, setting up value converters should be enough to get basic functionality.

**Comment by bricelam** [comment]
@xtellurian Does it have to be embedded JSON or will it work as a string containing JSON?

**Comment by bricelam** [comment]
Do we automatically embed JObject properties?
``` csharp
var jObject = new JObject();
new GeoJsonWriter.Write(point, new JTokenWriter(jObject));

var point = new GeoJsonReader().Read<Point>(new JTokenReader(jObject));
```

**Comment by AndriySvyryd** [comment]
We don't have any special handling for JObject, but string should work, I'll try it out later

**Comment by xtellurian** [comment]
Thanks for jumping on this to help :)

@bricelam 
I tried manually serializing the GeoJSON in CosmosDB and indexing with Azure Search (as below), and the search indexer seems like it works!

```json
{
"GeoLocation": "{ \"type\": \"Point\", \"coordinates\": [100.2093,-15.868]}"
}
```

@AndriySvyryd 
It sounds like I can use [value conversions](https://docs.microsoft.com/en-us/ef/core/modeling/value-conversions) to serialize/deserialize the GEOJson with the EF CosmosDB provider.

I don't see a need to use NetTopologySuite.IO.GeoJSON ...

**Comment by bricelam** [comment]
lol, using NTS just to serialize points is definitely overkill. It’ll come in handy for more complex spatial scenarios though

**Comment by cwoolum** [comment]
The problem with the solution above is that the whole `GeoLocation` object is actually saved as a string instead of a JSON literal. This means that Cosmos Indexing won't work. Is there a way to use a value converter but the force EF Core to use the converted value verbatim?

It should be saved as 
```
{
    "GeoLocation": {
        "type": "Point", 
        "coordinates": [100.2093,-15.868]
    }
}
```

**Comment by AndriySvyryd** [comment]
@cwoolum You'd be able to do it when https://github.com/dotnet/efcore/issues/20584 is implemented.
For now you can [modify the underlying JObject directly](https://docs.microsoft.com/en-us/ef/core/providers/cosmos/unstructured-data)

**Comment by cwoolum** [comment]
I think that by utilizing that with a raw sql command, you could feasibly use `ST_DISTANCE` querying using EF Core and Cosmos.

**Comment by johnn82** [comment]
> ```json
> "GeoLocation": "{ \"type\": \"Point\", \"coordinates\": [100.2093,-15.868]}"
> ```

 @xtellurian this is literally not working! I just tried it with the data explorer. I added the GeoLocation property to a document and then filtered using this query: "WHERE ST_DISTANCE(c.GeoLocation, {"type": "Point", "coordinates":[100.209,-15.86]}) < 40000"
 
 As said by @cwoolum, it cannot work if the GeoJSON object is saved as a serialized string.
 
 Does anybody knows how to make it works?

**Comment by xtellurian** [comment]
@johnn82 It worked with [Azure Search](https://docs.microsoft.com/en-us/azure/search/search-query-odata-geo-spatial-functions) but I never tried using purely Cosmos

**Comment by WillTheProgrammer** [comment]
> erialize the GEOJson with the EF CosmosDB provider.

@xtellurian  - please, please, please provide a quick writeup and code sample of how to do this.

**Comment by ajcvickers** [comment]
This issue is in the Backlog milestone. We have not yet fully planned the next release but we will [re-assess the backlog](https://docs.microsoft.com/ef/core/what-is-new/release-planning) as part of this. However, keep in mind that there are many other high priority features with which it will be competing for resources. Make sure to vote (👍) for this issue if it is important to you.
